src/QsCompiler/QirGeneration/AddOutputRecording.cs

```diff
@@ -0,0 +1,572 @@
+        {
+            var apiNamespace =
+                namespaces.FirstOrDefault(x => x.Name == OutputRecorderDefinition.NamespaceName)
+                ?? new QsNamespace(
+                    OutputRecorderDefinition.NamespaceName,
+                    ImmutableArray<QsNamespaceElement>.Empty,
+                    Enumerable.Empty<ImmutableArray<string>>().ToLookup(x => default(string)));
+
+            var definedElements = apiNamespace.Elements.Partition(e => e.IsQsCallable);
+            var definedCallables = definedElements.Item1.ToDictionary(
+                e => e.GetFullName().Name,
+                e => ((QsNamespaceElement.QsCallable)e).Item);
+
+            var apiSource = new Source(Path.GetTempFileName(), QsNullable<string>.Null);
+            foreach (var recorder in recorderAPI)
+            {
+                var recorderDeclaration = definedCallables.TryGetValue(recorder.QSharpName.Name, out var decl) ? decl : CreateOutputRecorder(recorder, apiSource);
+                if (SymbolResolution.TryGetTargetInstructionName(recorderDeclaration.Attributes).IsNull)
+                {
+                    recorderDeclaration = recorderDeclaration.AddAttribute(
+                        AttributeUtils.BuildAttribute(
+                            BuiltIn.TargetInstruction.FullName,
+                            AttributeUtils.StringArgument(recorder.TargetInstructionName)));
+                }
+
+                definedCallables[recorder.QSharpName.Name] = recorderDeclaration;
+                if (!(recorderDeclaration.Kind.IsFunction
+                    && recorderDeclaration.Signature.ArgumentType.Equals(recorder.QSharpSignature.ArgumentType)
+                    && recorderDeclaration.Signature.ReturnType.Resolution.IsUnitType
+                    && SymbolResolution.TryGetTargetInstructionName(recorderDeclaration.Attributes) is var instr
+                    && instr.IsValue && instr.Item == recorder.TargetInstructionName))
+                {
+                    throw new InvalidOperationException("callable with the expected recorder name is already defined bu t doesn't have the expected signature");
+                }
+            }
+
+            apiNamespace = apiNamespace.WithElements(_ =>
+                definedElements.Item2.Concat(definedCallables.Values.Select(QsNamespaceElement.NewQsCallable))
+                .ToImmutableArray());
+            return namespaces.Where(x => x.Name != apiNamespace.Name).Append(apiNamespace).ToImmutableArray();
+        }
+
+        private static (SymbolTuple, ImmutableDictionary<string, LocalVariableDeclaration<string, ResolvedType>>) CreateDeconstruction(ResolvedType returnType, int enumerationStart = 0)
+        {
+            var newVars = ImmutableDictionary.CreateBuilder<string, LocalVariableDeclaration<string, ResolvedType>>();
+            static string MakeVariableName(int enumeration) =>
+                $"__rtrnVal{enumeration}__";
```

Comment:
I know it is only used here, but "rtrnVal" as a label should be defined somewhere more accessible, like as a static readonly string near the top of the file.

Reply:
I move it out of the function and made it a private static member instead.